enabler/src/de/schildbach/pte/VrsProvider.java

```diff
@@ -0,0 +1,904 @@
+		if (product.equals("LongDistanceTrains")) {
+			return Product.HIGH_SPEED_TRAIN;
+		} else if (product.equals("RegionalTrains")) {
+			return Product.REGIONAL_TRAIN;
+		} else if (product.equals("SuburbanTrains")) {
+			return Product.SUBURBAN_TRAIN;
+		} else if (product.equals("Underground") || product.equals("LightRail") && number.startsWith("U")) {
+			return Product.SUBWAY;
+		} else if (product.equals("LightRail")) {
+			return Product.TRAM;
+		} else if (product.equals("Bus") || product.equals("CommunityBus") || product.equals("RailReplacementServices")) {
+			return Product.BUS;
+		} else if (product.equals("Boat")) {
+			return Product.FERRY;
+		} else if (product.equals("OnDemandServices")) {
+			return Product.ON_DEMAND;
+		} else {
+			throw new IllegalArgumentException("unknown product: '" + product + "'");
+		}
+	}
+
+	private String generateProducts(Set<Product> products) {
+		StringBuilder ret = new StringBuilder();
+		Iterator<Product> it = products.iterator();
+		while (it.hasNext()) {
+			final Product product = it.next();
+			if (ret.length() > 0 && !ret.substring(ret.length() - 1).equals(",")) {
+				ret.append(",");
+			}
+			ret.append(generateProduct(product));
+		}
+		return ret.toString();
+	}
+
+	private String generateProduct(Product product) {
+		switch (product) {
+		case BUS:
+			return "Bus,CommunityBus";
+		case CABLECAR:
+			return "";
+		case FERRY:
+			return "Boat";
+		case HIGH_SPEED_TRAIN:
+			return "LongDistanceTrains";
+		case ON_DEMAND:
+			return "OnDemandServices";
+		case REGIONAL_TRAIN:
+			return "RegionalTrains";
+		case SUBURBAN_TRAIN:
+			return "SuburbanTrains";
+		case SUBWAY:
+			return "LightRail,Underground";
+		case TRAM:
+			return "LightRail";
+		}
+		throw new IllegalArgumentException("unknown product: '" + product + "'");
+	}
+
+	public LocationWithPosition parseLocationAndPosition(JSONObject location) throws JSONException {
+		final LocationType locationType;
+		String id = null;
+		String name = null;
+		String position = null;
+		if (location.has("id")) {
+			locationType = LocationType.STATION;
+			id = location.getString("id");
+			name = location.getString("name");
+			// System.out.println("name: " + name);
+			for (Pattern pattern : nameWithPositionPatterns) {
+				Matcher matcher = pattern.matcher(name);
+				if (matcher.matches()) {
+					name = matcher.group(1);
+					position = matcher.group(2);
+					// System.out.println("MATCH! [" + name + "] [" + position.substring(position.lastIndexOf(" ") + 1) + "]");
+					break;
+				}
+			}
+		} else if (location.has("street")) {
+			locationType = LocationType.ADDRESS;
+			id = location.getString("tempId");
+			name = (location.getString("street") + " " + location.getString("number")).trim();
+		} else if (location.has("name")) {
+			locationType = LocationType.POI;
+			id = location.getString("tempId");
+			name = location.getString("name");
+		} else {
+			locationType = LocationType.ANY;
```

Comment:
It's maybe better to throw an exception here? Looks like an unexpected case to me.